### DISCUSSION TITLE
Get X coordinate from Y ???

### DISCUSSION THREAD

**Discussion by rs78GT** [bitcoin-core/secp256k1#1432] [Q&A] [answered]
Hi,
First of all, excuse my language but I am really tired of this.. could someone explain me, what is the RIGHT equation?
I am completely lost on this one.
What is the reason, to leave out half of the equation everywhere, what is about the secp256k1 mathematics on the Net?

The function is simple: y^2 = x^3 +7 (a = 0, b = 7,  so more than simple)
If I solve this one with GMP -MPZ.. the result is light years from the solution.
Oh.. might be there are some [MOD P]-s, but where.. everywhere. Oh.. maybe something on power of something.. (P-1) or (P+1).. not a big deal!

Sorry gentlemen, but could someone explain me how can I get the Y coordinate across  a  concrete example?

So.. for example: 

Private key:
0x7777777777777777777777777777777777777777777777777777777777777777 
X-Y coordinates:
mpz_init_set_str(x, "7962D45B38E8BCF82FA8EFA8432A01F20C9A53E24C7D3F11DF197CB8E70926DA", 16);
mpz_init_set_str(y, "7A3EF3EBAFC756DC3B24B75292D4CC5D71B170E97044A9858353443A96BAED23", 16);
						
mpz_pow_ui(y, y, 2);
mpz_mod(y, y, p);
gmp_printf("%Z02X\n", y1);
			
mpz_pow_ui(x1, x1, 3);
mpz_add_ui(x1, x1, 7);
mpz_mod(x1, x1, p);
gmp_printf("%Z02X\n", x1);

Result:
51A8F9D55FA628BD9E8B970CDA3C19004C1ABFFAB83CB65E48B837A9509A1366
51A8F9D55FA628BD9E8B970CDA3C19004C1ABFFAB83CB65E48B837A9509A1366

At least they are equal, but .. I tried with every kind of SQR, INVERT woodo.. not even close to Y.
So.. how the Hell is possible to get:
X = 7962D45B38E8BCF82FA8EFA8432A01F20C9A53E24C7D3F11DF197CB8E70926DA

(Sorry, If I messed up something..)
Thanks in advance!

**Comment by sipa** [comment]
To get the X coordinate from the Y coordinate, you need to solve *x<sup>3</sup> = y<sup>2</sup> - 7 (mod p)*, or in other words, find the modular cube roots of *y<sup>2</sup> - 7*. Due to the fact that *p mod 3 = 1*, this is particularly easy: *x = (y<sup>2</sup> - 7)<sup>(p + 2)/9</sup> (mod p)* to find one of the cube roots. They always come in triples (for any given Y coordinate on the curve there are 3 distinct X coordinates); the other two X's can be found by multiplying with the cube roots of 1 (namely 0x851695d49a83f8ef919bb86153cbcb16630fb68aed0a766a3ec693d68e6afa40 and 0x7ae96a2b657c07106e64479eac3434e99cf0497512f58995c1396c28719501ee).

This seems mostly unrelated to libsecp256k1 though, as it has no functionality to compute X from Y. There may be better places to ask about this (In fact, I recently answered a very related question here: https://bitcoin.stackexchange.com/a/120051/208).

**Comment by rs78GT** [comment]
Hi sipa,
good God.. are you never sleeping? :) I know, I know.. but I did not find anything on the GMP side. I try to find the way, to store only the Y coordinate, because probably the math "more cheap" than a lot of import, export. Now I try to understand what you just wrote.. for the first sight that is in chinese.. :)

**Reply by sipa** [reply, marked as answer]
I'm not sure what you're trying to do, but computing Y from X is a lot simpler than X from Y. That is what the compressed public key format is based on, which *is* supported by libsecp256k1.

For that, the relation is just *y = +- sqrt(x^3 + 7) mod p*.

**Comment by rs78GT** [comment]
That's what I thought! This is the reason, why I just wanna store the Y coordinate. No even or odd, like from X.. just straight point coordinates.
sipa, ppreciate your time,
Zsolt
